cmd/gogrep/main.go

```diff
@@ -21,44 +23,61 @@ func main() {
 	var before int
 	flag.IntVar(&before, "B", 0, "Shows number of lines before the Match.")
 
+	var recursive bool
+	flag.BoolVar(&recursive, "r", false, "Searches every file inside a directory.")
+
 	flag.Parse()
 
-	var fileContents []string
 	pattern := flag.Arg(0)
-	listOfFiles := iofile.ListFilesInDir(flag.Arg(1))
-	for i := 0; i < len(listOfFiles); i++ {
-		fileContents, _ = iofile.ReadFile(listOfFiles[i])
-		if caseSensitive {
-			output := gogrep.GrepCaseInsensitive(fileContents, pattern)
+	rootPath := flag.Arg(1)
 
-			for i := 0; i < len(output); i++ {
-				fmt.Println(output[i])
-			}
+	paths, err := iofile.ListFiles(rootPath)
+	if err != nil {
+		log.Fatal(err)
+	}
 
-		} else if count {
-			output := gogrep.GrepCount(fileContents, pattern)
-			fmt.Println(output)
+	outputChan := make(chan []string)
+	var wg sync.WaitGroup
+	for _, path := range paths {
+		wg.Add(1)
+		go worker(caseSensitive, count, after, before, path, pattern, outputChan, &wg)
+	}
 
-		} else if after > 0 {
-			output := gogrep.GrepAfter(after, fileContents, pattern)
+	go func() {
 
+		for output := range outputChan {
 			for i := 0; i < len(output); i++ {
 				fmt.Println(output[i])
```

Comment:
This is an impure function, I cannot test it. Better to push into a data structure and use outside of goroutines.

Reply:
Okay, Will create a separate function to WriteToStdout.